### PR TITLE
Retry requests that are probably safe to retry

### PR DESCRIPTION
**Pull request by tanakadesilva**

### CODE REVIEW

File: src/Cloud/PubSub/Http/Retry.hs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+retryDecider :: HttpT.RequestError -> Bool
+retryDecider = \case
+  HttpT.ResponseError status _ -> status `elem` retryStatuses
+  HttpT.DecodeError status _ _ -> status `elem` retryStatuses
+ where
+  retryStatuses =
+    [ Status.requestTimeout408
+    , Status.tooManyRequests429
+    , Status.serviceUnavailable503
+    , Status.gatewayTimeout504
```
Comment (StefanFehrenbach) [inline src/Cloud/PubSub/Http/Retry.hs:34]: Where does this list come from?

Compared to this list: https://cloud.google.com/pubsub/docs/reference/error-codes
We have 408, they do not.
They have 500, we do not.
They have 502, we do not.

File: src/Cloud/PubSub/Http/Retry.hs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  :: forall m a
+   . (MonadIO m, MonadLogger m)
+  => String
+  -> m (Either HttpT.RequestError a)
+  -> m (Either HttpT.RequestError a)
+httpClientRetry name action = Retry.retrying retryPolicy retryCheck mkAttempt
+ where
+  logAttempt :: Int -> m ()
+  logAttempt n =
+    $logError $ Text.pack $ mconcat ["retrying request", name, "(", show n, ")"]
```
Comment (StefanFehrenbach) [inline src/Cloud/PubSub/Http/Retry.hs:53]: We might want to add some spaces, otherwise: `retrying requestGET: some/path(42)`

Comment (tanakadesilva) [reply src/Cloud/PubSub/Http/Retry.hs]: Formatted it, thanks.

File: src/Cloud/PubSub/HttpClient.hs [hunk 3]
```diff
@@ -59,9 +63,11 @@ authedJsonGetRequest
...
   -> m r
 authedJsonGetRequest pathQueryParams = do
-  request  <- mkRequest "GET" pathQueryParams
-  response <- authJsonRequest request
-  return $ Http.getResponseBody response
+  request <- mkRequest "GET" pathQueryParams
+  (Retry.httpClientRetry name $ runRequest request) >>= either throwM return
+ where
+  name       = "GET: " <> HttpT.path pathQueryParams
+  runRequest = fmap decodeResponse . authJsonRequest
```
[the remark this answers is not in the document]
Comment (tanakadesilva) [reply src/Cloud/PubSub/HttpClient.hs:70]: Yes not using `decodeResponse` in `authedJsonGetRequest` was a bug, I think I wrote it before `decodeResponse` was written and forgot change it. 

Yes we should change `authedJsonGetRequest` type to return an either type to be consistent with the other functions and to make the the errors more explicit to the callee.

**Pull request merged**